### DISCUSSION TITLE
Unable to upload when using SSE-C - Error: <Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message>

### DISCUSSION THREAD

**Discussion by titiviking** [minio/minio#15021] [Q&A] [answered]
I'm attempting to perform a file upload using a presigned URL and applying SSE-C.

I'm able to obtain a correct presigned URL in the form of:
`https://files.mydomain.com/my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=202-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-Amz-Expires=7199&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220601%2Fus-east-1%2Fs3%2Faws4_request48c0516b45338cd72e2f70856d385790d20415d`

However, when performing the actual PUT of the file using this URL, I get the following error:
`<?xml version="1.0" encoding="UTF-8"?> <Error><Code>AccessDenied</Code><Message>There were headers present in the request which were not signed</Message><Key>file02.exe</Key><BucketName>my-bucket</BucketName><Resource>/my-bucket/file02.exe</Resource><RequestId>16F485EA6476D247</RequestId><HostId>9f9f0642-bb49-490d-b9f0-9e165a391ee7</HostId></Error>`

Both requests 'for presigned and for actual PUT), do embed the required headers to perform the upload:
```
GET /s3/getupload?name=file02.exe HTTP/1.1
> Host: api-local.mydomain.com
> User-Agent: curl/7.74.0
> Accept: */*
> X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
> X-Amz-Server-Side-Encryption-Customer-Key: PtYFz8gY7C42g5Z/ctPJjnagIhNjDN7pwkyM/MciD5o=
> X-Amz-Server-Side-Encryption-Customer-MD5: 51e42c8ed3e63d7f0bcc92e259324db9
```

The S3 Client that connects to Minio creates the PreSignedURL as follows:
```
Set<Entry<String, String>> entrySet = headers.entrySet();
for (Entry<String, String> e : entrySet) {
                    if(e.getKey().equals("x-amz-server-side-encryption-customer-key")) customerKey = e.getValue();
                    System.out.println("HTTP Header: " + e.getKey() + ": " + e.getValue());
}
// decode the base64 encoded string
byte[] decodedKey = Base64.getDecoder().decode(customerKey);
// rebuild key using SecretKeySpec
SecretKey originalKey = new SecretKeySpec(decodedKey, 0, decodedKey.length, "AES");

GeneratePresignedUrlRequest genreq = new GeneratePresignedUrlRequest(bucketName, filename, HttpMethod.PUT)
if(customerKey != null) {
       System.out.println("Encryption header present. Will request signed URL that corresponds to encryption.");
       genreq.setSSECustomerKey(new SSECustomerKey(originalKey));
}
genreq.setExpiration(new Date(System.currentTimeMillis() + 1000*60*120));
URL url = s3client.generatePresignedUrl(genreq);

```

Any idea what is going on ? Which header is being referred to ?

**Comment by harshavardhana** [comment]
Your presigned URL looks broken. Can you paste the exact URL that is generated?

**Comment by titiviking** [comment]
This should be better:
`https://files.mydomain.com/my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220601T191015Z&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-server-side-encryption-customer-key-md5&X-Amz-Expires=7199&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220601%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=2f11b36686609fb27b95467dd0986a3fbec1e241a90e8ce524a94220a1b1b145
`

_Note: I did modify the domain for confidentiality purposes..._

**Reply by harshavardhana** [reply]
> `https://files.mydomain.com/my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220601T191015Z&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-server-side-encryption-customer-key-md5&X-Amz-Expires=7199&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220601%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=2f11b36686609fb27b95467dd0986a3fbec1e241a90e8ce524a94220a1b1b145 `

So when you are uploading are you setting these values in your PUT request?

**Comment by titiviking** [comment]
See below:

**Obtained PreSigned URL**:
`https://files.mydomain.com/my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220602T064800Z&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-server-side-encryption-customer-key-md5&X-Amz-Expires=7200&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220602%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=bf0d0f3cd4b2c00c6f81bc75bc7c3cb9c07aede26c630f5af88fc06fb8431456`


**Using the URL in the PUT request**:
```
> PUT /my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220602T064800Z&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-server-side-encryption-customer-key-md5&X-Amz-Expires=7200&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220602%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=bf0d0f3cd4b2c00c6f81bc75bc7c3cb9c07aede26c630f5af88fc06fb8431456 HTTP/2
> Host: files.mydomain.com
> user-agent: curl/7.74.0
> accept: */*
> x-amz-server-side-encryption-customer-algorithm: AES256
> x-amz-server-side-encryption-customer-key: PtYFz8gY7C42g5Z/ctPJjnagIhNjDN7pwkyM/MciD5o=
> x-amz-server-side-encryption-customer-md5: 51e42c8ed3e63d7f0bcc92e259324db9
> content-length: 41022080
```

**Reply by harshavardhana** [reply]
You are not adding `X-Amz-Server-Side-Encryption-Customer-Key-Md5` instead your HTTP header has a typo `x-amz-server-side-encryption-customer-md5` 

Also `X-Amz-Server-Side-Encryption-Customer-Key-Md5` is base64 not hexadecimal value.

**Comment by harshavardhana** [comment]
You are not adding `X-Amz-Server-Side-Encryption-Customer-Key-Md5` instead your HTTP header has a typo `x-amz-server-side-encryption-customer-md5` 

Also `X-Amz-Server-Side-Encryption-Customer-Key-Md5` is base64 not hexadecimal value.

**Comment by titiviking** [comment]
I've made some progress. Thanks for highlighting that typo, I completely overlooked that one...
I've also attempted to base64 encode the MD5 Hex value, but without success. So I'm recreating all values from scratch.

I'm now facing a "Incorrect signature error"
`<?xml version="1.0" encoding="UTF-8"?> <Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><Key>file02.exe</Key><BucketName>my-bucket</BucketName><Resource>/my-bucket/file02.exe</Resource><RequestId>16F4C03A4946CE44</RequestId><HostId>9f9f0642-bb49-490d-b9f0-9e165a391ee7</HostId></Error>`


I created my AES keys as such:

```
$ dd if=/dev/urandom of=secret.key bs=32  count=1
$ openssl base64 -in secret.key
BMj+4MKG1cyDdGAxHsF7JzYBeqB1soJX81jgzF2CanQ=
$ md5sum -b < secret.key | openssl base64
YTlmNTczZmEzMmFjNzM4ZTY0NDNjOTUxNTBmYTJkYTQgKi0K
```


So there are the new header values for which I got the above error:

```
> x-amz-server-side-encryption-customer-algorithm: AES256
> x-amz-server-side-encryption-customer-key: BMj+4MKG1cyDdGAxHsF7JzYBeqB1soJX81jgzF2CanQ=
> x-amz-server-side-encryption-customer-key-md5: YTlmNTczZmEzMmFjNzM4ZTY0NDNjOTUxNTBmYTJkYTQgKi0K
```

**Reply by titiviking** [reply]
more info:

I've also attempted the commands described on Min.io documentation:
(https://docs.min.io/minio/baremetal/security/server-side-encryption/server-side-encryption-sse-c.html)

```
$ cat /dev/urandom | head -c 32 | base64 -
ubhLEH+ZnTSkrZ8wlvFMnwP7ytQhLXHQZNb6jJ6R7Ek=

// Digest calculation with https://v2.cryptii.com/base64/md5 using above value

$ echo 6363fe744f74ee8f280958ab2f185dde | base64
NjM2M2ZlNzQ0Zjc0ZWU4ZjI4MDk1OGFiMmYxODVkZGUK

```

I keep facing the issue....

**Reply by harshavardhana** [reply]
> I've made some progress. Thanks for highlighting that typo, I completely overlooked that one... I've also attempted to base64 encode the MD5 Hex value, but without success. So I'm recreating all values from scratch.
> 
> I'm now facing a "Incorrect signature error" `<?xml version="1.0" encoding="UTF-8"?> <Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><Key>file02.exe</Key><BucketName>my-bucket</BucketName><Resource>/my-bucket/file02.exe</Resource><RequestId>16F4C03A4946CE44</RequestId><HostId>9f9f0642-bb49-490d-b9f0-9e165a391ee7</HostId></Error>`
> 
> I created my AES keys as such:
> 
> ```
> $ dd if=/dev/urandom of=secret.key bs=32  count=1
> $ openssl base64 -in secret.key
> BMj+4MKG1cyDdGAxHsF7JzYBeqB1soJX81jgzF2CanQ=
> $ md5sum -b < secret.key | openssl base64
> YTlmNTczZmEzMmFjNzM4ZTY0NDNjOTUxNTBmYTJkYTQgKi0K
> ```
> 
> So there are the new header values for which I got the above error:
> 
> ```
> > x-amz-server-side-encryption-customer-algorithm: AES256
> > x-amz-server-side-encryption-customer-key: BMj+4MKG1cyDdGAxHsF7JzYBeqB1soJX81jgzF2CanQ=
> > x-amz-server-side-encryption-customer-key-md5: YTlmNTczZmEzMmFjNzM4ZTY0NDNjOTUxNTBmYTJkYTQgKi0K
> ```

Incorrect signature means your signature generation is wrong, you are mainly signing with a Host header but using a different Host header.

You need to figure this out on your end..

**Comment by titiviking** [comment]
My Host is configured everyware to be the same...

My Java client uses _https://files.mydomain.com_ which point to my reverse proxy that forwards to localhost:3000.
The customer side makes the PUT request to _https://files.mydomain.com_ using the presigned URL provided by the Java client.

So all domains/hosts are aligned...

Without encryption it works fine, so even if it signed the hosts... then encryption does not change the actual values being signed between requests with or without encryption...

**Reply by titiviking** [reply]
Doing an HTTP trace on Minio, I do confirm that the host/domain remains the same:

```
2022-06-02T13:53:57:000 [200 OK] s3.HeadBucket files.mydomain.com/my-bucket 172.17.0.8        657µs       ↑ 160 B ↓ 314 B
2022-06-02T13:53:58:000 [403 Forbidden] s3.PutObject files.mydomain.com/my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220602T135357Z&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-server-side-encryption-customer-key-md5&X-Amz-Expires=7199&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220602%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=797a01cd7108507f9c30be3980f8af2ae9b9308692b49038f2dc3baf8caf36dd  34.70.17.11       962µs       ↑ 229 B ↓ 732 B
```

**Reply by titiviking** [reply]
Other debug info (with more headers):

```
files.mydomain.com [REQUEST s3.PutObject] [2022-06-02T13:58:39:000] [Client IP: 34.7v.wx.yz]
files.mydomain.com PUT /my-bucket/file02.exe?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20220602T135838Z&X-Amz-SignedHeaders=host%3Bx-amz-server-side-encryption-customer-algorithm%3Bx-amz-server-side-encryption-customer-key%3Bx-amz-server-side-encryption-customer-key-md5&X-Amz-Expires=7199&X-Amz-Credential=BGBKXKS1CH27RC134U97%2F20220602%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=0721b6d8828dd0e8e0114682b1d94234b584655f1a2f4fa50f71bd40dbaf0e6d
files.mydomain.com Proto: HTTP/1.1
files.mydomain.com Host: files.mydomain.com
files.mydomain.com Connection: close
files.mydomain.com Content-Length: 41022080
files.mydomain.com User-Agent: curl/7.74.0
files.mydomain.com X-Forwarded-Proto: https
files.mydomain.com Accept: */*
files.mydomain.com X-Amz-Server-Side-Encryption-Customer-Algorithm: AES256
files.mydomain.com X-Amz-Server-Side-Encryption-Customer-Key: 1Z5Gk19ZHaBnEDYY38t+5ELi4psS88yYSeiRjMBNuz4=
files.mydomain.com X-Amz-Server-Side-Encryption-Customer-Key-Md5: NGYyMGRjMjUyMjk2Y2E3ODIyYjkwMDAwNjExZmY1MzYK
files.mydomain.com X-Forwarded-For: 34.7v.wx.yz
files.mydomain.com X-Real-Ip: 34.7v.wx.yz
files.mydomain.com <BODY>
files.mydomain.com [RESPONSE] [2022-06-02T13:58:39:000] [ Duration 1.166ms  ↑ 229 B  ↓ 732 B ]
files.mydomain.com 403 Forbidden
files.mydomain.com Accept-Ranges: bytes
files.mydomain.com Content-Type: application/xml
files.mydomain.com Server: MinIO
files.mydomain.com X-Content-Type-Options: nosniff
files.mydomain.com X-Xss-Protection: 1; mode=block
files.mydomain.com Content-Length: 409
files.mydomain.com Content-Security-Policy: block-all-mixed-content
files.mydomain.com Strict-Transport-Security: max-age=31536000; includeSubDomains
files.mydomain.com Vary: Origin,Accept-Encoding
files.mydomain.com X-Amz-Request-Id: 16F4D26ADE3DDC13
files.mydomain.com <?xml version="1.0" encoding="UTF-8"?>
<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><Key>file02.exe</Key><BucketName>my-bucket</BucketName><Resource>/my-bucket/file02.exe</Resource><RequestId>16F4D26ADE3DDC13</RequestId><HostId>9f9f0642-bb49-490d-b9f0-9e165a391ee7</HostId></Error>
```

**Comment by titiviking** [comment, marked as answer]
Issue identified ;-)
The commands used to calculate the Base64 encoded MD5 of the key were incorrect... the `md5sum` only allows to obtain a hex encoded md5... however the Base64 should be calculated on the binary value of the MD5...

I used the following command to obtain correct Base64 encoded md5:
`openssl dgst -md5 -binary hmac.key | base64`

Putting this new value as header value the upload was successful.

**Reply by titiviking** [reply]
Full set of commands to obtain the required values:

```
$ dd if=/dev/urandom of=secret.key bs=32  count=1    // Generate 256-bit key and store it in secret.key
$ openssl base64 -in secret.key   
1Z5Gk19ZHaBnEDYY38t+5ELi4psS88yYSeiRjMBNuz4=        // Base64 encoded representation of the key
$ openssl dgst -md5 -binary secret.key | base64
TyDcJSKWyngiuQAAYR/1Ng==                             // Base64 encoded MD5 digest of the key
```
